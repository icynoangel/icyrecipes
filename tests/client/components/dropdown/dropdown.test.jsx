import React from 'react';
import ReactDOM from 'react-dom';
import {mount} from 'enzyme';
import Dropdown from './../../../../src/client/js/components/dropdown/dropdown';
import renderer from 'react-test-renderer';
import enzymeToJson from 'enzyme-to-json';

describe('<Dropdown />', function() {
  this.getComponent = (items = this.items) => {
    return (
      <Dropdown
        onChange={this.onChange}
        placeholder={this.placeholder}
        items={items}
      />
    );
  };

  this.items = [
    {
      label: 'Select option 1',
      value: 'opt_1'
    },
    {
      label: 'Select option 2',
      value: 'opt_2'
    },
    {
      label: 'Select option 3',
      value: 'opt_3'
    },
    {
      label: 'Select option 4',
      value: 'opt_4'
    }
  ];

  beforeEach(() => {
    this.onChange = jest.fn();
    this.placeholder = 'Custom select';
  });

  it('Should render closed <Dropdown />', () => {
    const dropdown = renderer.create(this.getComponent()).toJSON();
    expect(dropdown).toMatchSnapshot();
  });

  it('Should render <Dropdown /> with default selected value', () => {
    const items = this.items.slice(0);
    items.push({
      label: 'Select option 5',
      value: 'opt_5',
      selected: true
    });
    const wrapper = mount(this.getComponent(items));
    wrapper.find('.dropdown__control').simulate('click');

    expect(enzymeToJson(wrapper)).toMatchSnapshot();
  });

  it('Should not call onChange for current selection', () => {
    const items = this.items.slice(0);
    items.push({
      label: 'Select option 5',
      value: 'opt_5',
      selected: true
    });
    const wrapper = mount(this.getComponent(items));
    wrapper.find('.dropdown__control').simulate('click');

    wrapper
      .find('.dropdown__content__option')
      .at(4)
      .simulate('click');

    expect(this.onChange).not.toHaveBeenCalled();
  });

  it('Should open dropdown', () => {
    const wrapper = mount(this.getComponent());
    wrapper.find('.dropdown__control').simulate('click');

    expect(enzymeToJson(wrapper)).toMatchSnapshot();
  });

  it('Should select a value and call onChange', () => {
    const wrapper = mount(this.getComponent());
    wrapper.find('.dropdown__control').simulate('click');

    wrapper
      .find('.dropdown__content__option')
      .at(2)
      .simulate('click');

    expect(enzymeToJson(wrapper)).toMatchSnapshot();
  });

  it('Should addEventListeners when mounting and removeEventListenrs when unmounting', () => {
    const addEventsSpy = jest.spyOn(Dropdown.prototype, 'addListeners');
    const removeEventsSpy = jest.spyOn(Dropdown.prototype, 'removeListeners');
    const wrapper = mount(this.getComponent());

    expect(addEventsSpy).toHaveBeenCalled();
    wrapper.unmount();
    expect(removeEventsSpy).toHaveBeenCalled();
  });

  it('Should close dropdown on document click', () => {
    const wrapper = mount(this.getComponent());
    wrapper.find('.dropdown__control').simulate('click');

    wrapper.instance().handleDocumentClick();

    expect(wrapper.state('isOpen')).toBe(false);
  });
});
