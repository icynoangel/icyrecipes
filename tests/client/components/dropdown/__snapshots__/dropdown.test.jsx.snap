// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dropdown /> Should open dropdown 1`] = `
<Dropdown
  items={
    Array [
      Object {
        "label": "Select option 1",
        "value": "opt_1",
      },
      Object {
        "label": "Select option 2",
        "value": "opt_2",
      },
      Object {
        "label": "Select option 3",
        "value": "opt_3",
      },
      Object {
        "label": "Select option 4",
        "value": "opt_4",
      },
    ]
  }
  onChange={[MockFunction]}
  placeholder="Custom select"
  type="medium"
  width=""
>
  <div
    className="dropdown --open --medium"
  >
    <div
      className="dropdown__control"
      onClick={[Function]}
    >
      <div
        className="dropdown__control__select"
      >
        Custom select
      </div>
      <div
        className="dropdown__control__arrow"
      />
    </div>
    <div
      className="dropdown__content"
    >
      <div
        className="dropdown__content__option"
        key="dropdown-option-opt_1"
        onClick={[Function]}
      >
        Select option 1
      </div>
      <div
        className="dropdown__content__option"
        key="dropdown-option-opt_2"
        onClick={[Function]}
      >
        Select option 2
      </div>
      <div
        className="dropdown__content__option"
        key="dropdown-option-opt_3"
        onClick={[Function]}
      >
        Select option 3
      </div>
      <div
        className="dropdown__content__option"
        key="dropdown-option-opt_4"
        onClick={[Function]}
      >
        Select option 4
      </div>
    </div>
  </div>
</Dropdown>
`;

exports[`<Dropdown /> Should render <Dropdown /> with default selected value 1`] = `
<Dropdown
  items={
    Array [
      Object {
        "label": "Select option 1",
        "value": "opt_1",
      },
      Object {
        "label": "Select option 2",
        "value": "opt_2",
      },
      Object {
        "label": "Select option 3",
        "value": "opt_3",
      },
      Object {
        "label": "Select option 4",
        "value": "opt_4",
      },
      Object {
        "label": "Select option 5",
        "selected": true,
        "value": "opt_5",
      },
    ]
  }
  onChange={[MockFunction]}
  placeholder="Custom select"
  type="medium"
  width=""
>
  <div
    className="dropdown --open --medium"
  >
    <div
      className="dropdown__control"
      onClick={[Function]}
    >
      <div
        className="dropdown__control__select"
      >
        Select option 5
      </div>
      <div
        className="dropdown__control__arrow"
      />
    </div>
    <div
      className="dropdown__content"
    >
      <div
        className="dropdown__content__option"
        key="dropdown-option-opt_1"
        onClick={[Function]}
      >
        Select option 1
      </div>
      <div
        className="dropdown__content__option"
        key="dropdown-option-opt_2"
        onClick={[Function]}
      >
        Select option 2
      </div>
      <div
        className="dropdown__content__option"
        key="dropdown-option-opt_3"
        onClick={[Function]}
      >
        Select option 3
      </div>
      <div
        className="dropdown__content__option"
        key="dropdown-option-opt_4"
        onClick={[Function]}
      >
        Select option 4
      </div>
      <div
        className="dropdown__content__option --selected"
        key="dropdown-option-opt_5"
        onClick={[Function]}
      >
        Select option 5
      </div>
    </div>
  </div>
</Dropdown>
`;

exports[`<Dropdown /> Should render auto <Dropdown /> 1`] = `
<Dropdown
  items={
    Array [
      Object {
        "label": "Select option 1",
        "value": "opt_1",
      },
      Object {
        "label": "Select option 2",
        "value": "opt_2",
      },
      Object {
        "label": "Select option 3",
        "value": "opt_3",
      },
      Object {
        "label": "Select option 4",
        "value": "opt_4",
      },
    ]
  }
  onChange={[MockFunction]}
  placeholder="Custom select"
  type="medium"
  width="auto"
>
  <div
    className="dropdown --medium --auto"
  >
    <div
      className="dropdown__control"
      onClick={[Function]}
    >
      <div
        className="dropdown__control__select"
      >
        Custom select
      </div>
      <div
        className="dropdown__control__arrow"
      />
    </div>
  </div>
</Dropdown>
`;

exports[`<Dropdown /> Should render closed <Dropdown /> 1`] = `
<Dropdown
  items={
    Array [
      Object {
        "label": "Select option 1",
        "value": "opt_1",
      },
      Object {
        "label": "Select option 2",
        "value": "opt_2",
      },
      Object {
        "label": "Select option 3",
        "value": "opt_3",
      },
      Object {
        "label": "Select option 4",
        "value": "opt_4",
      },
    ]
  }
  onChange={[MockFunction]}
  placeholder="Custom select"
  type="medium"
  width=""
>
  <div
    className="dropdown --medium"
  >
    <div
      className="dropdown__control"
      onClick={[Function]}
    >
      <div
        className="dropdown__control__select"
      >
        Custom select
      </div>
      <div
        className="dropdown__control__arrow"
      />
    </div>
  </div>
</Dropdown>
`;

exports[`<Dropdown /> Should render tiny <Dropdown /> 1`] = `
<Dropdown
  items={
    Array [
      Object {
        "label": "Select option 1",
        "value": "opt_1",
      },
      Object {
        "label": "Select option 2",
        "value": "opt_2",
      },
      Object {
        "label": "Select option 3",
        "value": "opt_3",
      },
      Object {
        "label": "Select option 4",
        "value": "opt_4",
      },
    ]
  }
  onChange={[MockFunction]}
  placeholder="Custom select"
  type="tiny"
  width=""
>
  <div
    className="dropdown --tiny"
  >
    <div
      className="dropdown__control"
      onClick={[Function]}
    >
      <div
        className="dropdown__control__select"
      >
        Custom select
      </div>
      <div
        className="dropdown__control__arrow"
      />
    </div>
  </div>
</Dropdown>
`;

exports[`<Dropdown /> Should select a value and call onChange 1`] = `
<Dropdown
  items={
    Array [
      Object {
        "label": "Select option 1",
        "value": "opt_1",
      },
      Object {
        "label": "Select option 2",
        "value": "opt_2",
      },
      Object {
        "label": "Select option 3",
        "value": "opt_3",
      },
      Object {
        "label": "Select option 4",
        "value": "opt_4",
      },
    ]
  }
  onChange={
    [MockFunction] {
      "calls": Array [
        Array [
          "opt_3",
        ],
      ],
    }
  }
  placeholder="Custom select"
  type="medium"
  width=""
>
  <div
    className="dropdown --medium"
  >
    <div
      className="dropdown__control"
      onClick={[Function]}
    >
      <div
        className="dropdown__control__select"
      >
        Select option 3
      </div>
      <div
        className="dropdown__control__arrow"
      />
    </div>
  </div>
</Dropdown>
`;
